addons/controllers/cpi/cpiconfig_utils.go

```diff
@@ -0,0 +1,307 @@
+			clusterName = ownerRef.Name
+			break
+		}
+	}
+	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: cpiConfig.Namespace, Name: clusterName}, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info(fmt.Sprintf("Cluster resource '%s/%s' not found", cpiConfig.Namespace, clusterName))
+		}
+		r.Log.Error(err, fmt.Sprintf("Unable to fetch cluster '%s/%s'", cpiConfig.Namespace, clusterName))
+		return nil, err
+	}
+	r.Log.Info(fmt.Sprintf("Cluster resource '%s/%s' is successfully found", cpiConfig.Namespace, clusterName))
+	return cluster, nil
+}
+
+// getVSphereCluster gets the VSphereCluster CR for the cluster object
+func (r *CPIConfigReconciler) getVSphereCluster(ctx context.Context, cluster *clusterapiv1beta1.Cluster) (*capvv1beta1.VSphereCluster, error) {
+	vsphereCluster := &capvv1beta1.VSphereCluster{}
+	if err := r.Client.Get(ctx, types.NamespacedName{
+		Namespace: cluster.Namespace,
+		Name:      cluster.Name,
+	}, vsphereCluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, errors.Errorf("VSphereCluster %s/%s not found", cluster.Namespace, cluster.Name)
+		}
+		return nil, errors.Errorf("VSphereCluster %s/%s could not be fetched, error %v", cluster.Namespace, cluster.Name, err)
+	}
+	return vsphereCluster, nil
+}
+
+func getSecret(ctx context.Context, c client.Client, namespace, name string) (*v1.Secret, error) {
+	secret := &v1.Secret{}
+	if err := c.Get(ctx, types.NamespacedName{Namespace: namespace, Name: name}, secret); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, errors.Errorf("Secret %s/%s not found", namespace, name)
+		}
+		return nil, errors.Errorf("Secret %s/%s could not be fetched, error %v", namespace, name, err)
+	}
+	return secret, nil
+}
+
+func getUsernameAndPasswordFromSecret(_ context.Context, _ client.Client, s *v1.Secret) (string, string, error) {
+	username, exists := s.Data["username"]
+	if !exists {
+		return "", "", errors.Errorf("Secret %s/%s doesn't have string data with username", s.Namespace, s.Name)
+	}
+	password, exists := s.Data["password"]
+	if !exists {
+		return "", "", errors.Errorf("Secret %s/%s doesn't have string data with password", s.Namespace, s.Name)
+	}
+	return string(username), string(password), nil
+}
+
+// controlPlaneName returns the control plane name for a cluster name
+func controlPlaneName(clusterName string) string {
+	return fmt.Sprintf("%s-control-plane", clusterName)
+}
+
+// tryParseClusterVariableBool tries to parse a boolean cluster variable,
+// info any error that occurs
+func (r *CPIConfigReconciler) tryParseClusterVariableBool(_ context.Context, cluster *clusterapiv1beta1.Cluster, variableName string) bool {
+	res, err := util.ParseClusterVariableBool(cluster, variableName)
+	if err != nil {
+		r.Log.Info(fmt.Sprintf("cannot parse cluster variable with key %s", variableName))
+	}
+	return res
+}
+
+// tryParseClusterVariableString tries to parse a string cluster variable,
+// info any error that occurs
```

Comment:
Why do we need to have a separate ```tryParseClusterVariableString()``` function where we could directly call ```util.ParseClusterVariableString()```?

Reply:
don't wanna make the function bloated with too many `if ... then print`. let's keep it this way.